File: app/schemas/validation.py
"""Request and response bodies for the validation screens (task 103, SPEC §6.6).

Two families. The **audit** responses carry a report the worker stored as JSONB and typed
here for the client: the chunking report's histogram, distribution and findings, the
embedding report's checks. The **evaluation** bodies are sets, items and runs — and a run's
per-item results ride on the run row as the worker wrote them, typed here the same way.

An item's label in a request is the minimum a person or Try retrieval has in hand: a chunk
id and its document. The service fills in the text and the name from the index, because
the text is what makes the label survive a reindex and the client should not be trusted to
copy it faithfully.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.db.models.validation import MAX_QUESTION_LENGTH, MAX_SET_NAME_LENGTH

if TYPE_CHECKING:
    from app.db.models import EvaluationItem, EvaluationRun, EvaluationSet, IndexAudit
    from app.services.evaluation_service import (
        GenerateResult,
        ImportResult,
        RunDiff,
        SetDetail,
        SetView,
    )
    from app.services.evaluation_store import ItemCounts
    from app.services.index_auditor import AuditAlert, AuditStatus


# ---------------------------------------------------------------------------
# audits
# ---------------------------------------------------------------------------

AuditKind = Literal["chunking", "embedding"]
Severity = Literal["green", "amber", "red"]


class FindingDocumentResponse(BaseModel):
    id: str
    source_name: str
    count: int


class FindingResponse(BaseModel):
    code: str
    severity: Severity
    title: str
    count: int
    detail: str
    documents: list[FindingDocumentResponse] = Field(default_factory=list)
    document_count: int = 0
    action: Literal["compare", "reindex", "platform", "none"] = "none"


class DistributionResponse(BaseModel):
    chunks: int
    min_tokens: int
    median_tokens: int
    p95_tokens: int
    max_tokens: int
    at_ceiling: int
    mid_sentence: int


class HistogramBucketResponse(BaseModel):
    lower: int
    upper: int
    count: int


class HistogramResponse(BaseModel):
    bucket_tokens: int
    buckets: list[HistogramBucketResponse]


class FormatReportResponse(BaseModel):
    kind: str
    points: int
    documents: int
    chunk_size: int
    distribution: DistributionResponse
    histogram: HistogramResponse
    findings: list[FindingResponse]


class ChunkingReportResponse(BaseModel):
    kind: Literal["chunking"] = "chunking"
    points: int
    summary_points: int = 0
    documents: int
    documents_without_points: int = 0
    chunk_size: int
    distribution: DistributionResponse
    histogram: HistogramResponse
    formats: list[FormatReportResponse]
    findings: list[FindingResponse]
    fingerprints: dict[str, int] = Field(default_factory=dict)
    severity: Severity = "green"


class NormsResponse(BaseModel):
    min: float
    median: float
    p95: float
    max: float


class AgreementResponse(BaseModel):
    sampled: int
    agreed: int
    rate: float | None
    cross_document_similarity: float | None
    worst: list[FindingDocumentResponse]


class DriftResponse(BaseModel):
    sampled: int
    mean: float
    min: float
    p5: float
    below: float
    shape: Literal["healthy", "bimodal", "offset", "low"]
    model: str


class EmbeddingReportResponse(BaseModel):
    kind: Literal["embedding"] = "embedding"
    points: int
    scanned: int
    expected_dimension: int
    dimensions: dict[str, int]
    expected_model: str
    document_models: dict[str, int]
    norms: NormsResponse | None
    zero_vectors: int
    identical_vectors: int
    agreement: AgreementResponse | None
    drift: DriftResponse | None
    findings: list[FindingResponse]
    severity: Severity = "green"


class AuditResponse(BaseModel):
    id: uuid.UUID
    connector_id: uuid.UUID
    kind: AuditKind
    status: Literal["running", "succeeded", "failed"]
    created_at: datetime
    finished_at: datetime | None
    points: int
    drift_sample: int | None
    severity: Severity | None
    error: str | None
    report: ChunkingReportResponse | EmbeddingReportResponse | None

    @classmethod
    def of(cls, audit: IndexAudit) -> AuditResponse:
        report: ChunkingReportResponse | EmbeddingReportResponse | None = None
        if audit.status == "succeeded" and audit.report:
            if audit.kind == "chunking":
                report = ChunkingReportResponse.model_validate(audit.report)
            else:
                report = EmbeddingReportResponse.model_validate(audit.report)
        return cls(
            id=audit.id,
            connector_id=audit.connector_id,
            kind=audit.kind,  # type: ignore[arg-type]
            status=audit.status,  # type: ignore[arg-type]
            created_at=audit.created_at,
            finished_at=audit.finished_at,
            points=audit.points,
            drift_sample=audit.drift_sample,
            severity=audit.severity,  # type: ignore[arg-type]
            error=audit.error,
            report=report,
        )


class DriftEstimateResponse(BaseModel):
    points: int
    sample: int
    tokens: int


class AuditStatusResponse(BaseModel):
    """The connector's Validation section: the latest of each kind, and the price of the
    one check that spends."""

    chunking: AuditResponse | None
    embedding: AuditResponse | None
    drift_estimate: DriftEstimateResponse

    @classmethod
    def of(cls, status: AuditStatus) -> AuditStatusResponse:
        return cls(
            chunking=AuditResponse.of(status.chunking) if status.chunking else None,
            embedding=AuditResponse.of(status.embedding) if status.embedding else None,
            drift_estimate=DriftEstimateResponse(
                points=status.drift_estimate.points,
                sample=status.drift_estimate.sample,
                tokens=status.drift_estimate.tokens,
            ),
        )


class AuditRequest(BaseModel):
    """``drift_sample`` is the embedding audit's one spending option: how many chunks to
    re-embed with the current model. Omit it and the audit reads only."""

    model_config = ConfigDict(extra="forbid")

    drift_sample: int | None = Field(default=None, ge=1, le=1000)


class AuditAlertResponse(BaseModel):
    connector_id: uuid.UUID
    connector_name: str | None
    kind: AuditKind
    audit_id: uuid.UUID
    finding: str
    created_at: datetime

    @classmethod
    def of(cls, alert: AuditAlert) -> AuditAlertResponse:
        return cls(
            connector_id=alert.connector_id,
            connector_name=alert.connector_name,
            kind=alert.kind,  # type: ignore[arg-type]
            audit_id=alert.audit_id,
            finding=alert.finding,
            created_at=alert.created_at,
        )


class AuditAlertList(BaseModel):
    items: list[AuditAlertResponse]


# ---------------------------------------------------------------------------
# evaluation sets and items
# ---------------------------------------------------------------------------

ItemSource = Literal["log", "citation", "manual", "generated"]


class LabelRequest(BaseModel):
    """A chunk somebody says answers the question. The text is filled in from the index."""

    model_config = ConfigDict(extra="forbid")

    chunk_id: str = Field(min_length=1, max_length=64)
    document_id: uuid.UUID


class LabelResponse(BaseModel):
    chunk_id: str
    document_id: str | None
    source_name: str | None
    text: str | None


class ItemCountsResponse(BaseModel):
    total: int = 0
    verified: int = 0
    generated: int = 0
    negatives: int = 0

    @classmethod
    def of(cls, counts: ItemCounts) -> ItemCountsResponse:
        return cls(
            total=counts.total,
            verified=counts.verified,
            generated=counts.generated,
            negatives=counts.negatives,
        )


class EvaluationItemResponse(BaseModel):
    id: uuid.UUID
    set_id: uuid.UUID
    question: str
    relevant: list[LabelResponse]
    relevant_document_ids: list[str]
    source: ItemSource
    verified: bool
    #: No label at all: a question the corpus should answer with nothing.
    negative: bool
    notes: str | None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def of(cls, item: EvaluationItem) -> EvaluationItemResponse:
        labels = [
            LabelResponse(
                chunk_id=str(entry.get("chunk_id", "")),
                document_id=_optional(entry.get("document_id")),
                source_name=_optional(entry.get("source_name")),
                text=_optional(entry.get("text")),
            )
            for entry in (item.relevant or [])
            if isinstance(entry, dict) and entry.get("chunk_id")
        ]
        documents = [str(value) for value in (item.relevant_document_ids or [])]
        return cls(
            id=item.id,
            set_id=item.set_id,
            question=item.question,
            relevant=labels,
            relevant_document_ids=documents,
            source=item.source,  # type: ignore[arg-type]
            verified=item.verified,
            negative=not labels and not documents,
            notes=item.notes,
            created_at=item.created_at,
            updated_at=item.updated_at,
        )


class EvaluationItemRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    question: str = Field(min_length=1, max_length=MAX_QUESTION_LENGTH)
    relevant: list[LabelRequest] = Field(default_factory=list, max_length=50)
    relevant_document_ids: list[uuid.UUID] = Field(default_factory=list, max_length=50)
    #: A person adding an item has checked it; ``manual`` is verified unless said otherwise.
    verified: bool = True
    notes: str | None = Field(default=None, max_length=2000)


class EvaluationItemPatch(BaseModel):
    """Every field optional; ``relevant: []`` with ``relevant_document_ids: []`` makes the
    item a negative."""

    model_config = ConfigDict(extra="forbid")

    question: str | None = Field(default=None, min_length=1, max_length=MAX_QUESTION_LENGTH)
    relevant: list[LabelRequest] | None = Field(default=None, max_length=50)
    relevant_document_ids: list[uuid.UUID] | None = Field(default=None, max_length=50)
    verified: bool | None = None
    notes: str | None = Field(default=None, max_length=2000)

    def patch(self) -> dict[str, Any]:
        """The fields that were sent. ``notes: null`` clears the notes; a null anywhere
        else is not a value and is dropped."""
        values = self.model_dump(exclude_unset=True)
        patch = {key: value for key, value in values.items() if value is not None or key == "notes"}
        if patch.get("relevant") is not None:
            patch["relevant"] = [
                {"chunk_id": label["chunk_id"], "document_id": str(label["document_id"])}
                for label in patch["relevant"]
            ]
        if patch.get("relevant_document_ids") is not None:
            patch["relevant_document_ids"] = [str(v) for v in patch["relevant_document_ids"]]
        return patch


class EvaluationRunSummaryResponse(BaseModel):
    """A run as the history table shows it: status, progress, the headline numbers."""

    id: uuid.UUID
    set_id: uuid.UUID
    status: Literal["queued", "running", "succeeded", "failed"]
    created_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    total_items: int
    completed_items: int
    #: The unsaved patch the run applied, or ``None`` for the saved gateway.
    patch: dict[str, Any] | None
    metrics: dict[str, Any]
    snapshot: dict[str, Any]
    config: dict[str, Any]
    error: str | None

    @classmethod
    def of(cls, run: EvaluationRun) -> EvaluationRunSummaryResponse:
        return cls(
            id=run.id,
            set_id=run.set_id,
            status=run.status,  # type: ignore[arg-type]
            created_at=run.created_at,
            started_at=run.started_at,
            finished_at=run.finished_at,
            total_items=run.total_items,
            completed_items=run.completed_items,
            patch=dict(run.patch) if run.patch else None,
            metrics=dict(run.metrics or {}),
            snapshot=dict(run.snapshot or {}),
            config=dict(run.config or {}),
            error=run.error,
        )


class EvaluationRunResponse(EvaluationRunSummaryResponse):
    """The whole run: the summary plus every item's result."""

    results: list[dict[str, Any]]

    @classmethod
    def of(cls, run: EvaluationRun) -> EvaluationRunResponse:
        summary = EvaluationRunSummaryResponse.of(run)
        return cls(**summary.model_dump(), results=[dict(r) for r in (run.results or [])])


class EvaluationSetResponse(BaseModel):
    id: uuid.UUID
    gateway_id: uuid.UUID
    name: str
    description: str | None
    created_at: datetime
    updated_at: datetime
    counts: ItemCountsResponse
    last_run: EvaluationRunSummaryResponse | None = None

    @classmethod
    def of(cls, view: SetView) -> EvaluationSetResponse:
        return cls(
            **_set_fields(view.set),
            counts=ItemCountsResponse.of(view.counts),
            last_run=EvaluationRunSummaryResponse.of(view.last_run) if view.last_run else None,
        )


class EvaluationSetDetailResponse(BaseModel):
    id: uuid.UUID
    gateway_id: uuid.UUID
    name: str
    description: str | None
    created_at: datetime
    updated_at: datetime
    counts: ItemCountsResponse
    items: list[EvaluationItemResponse]

    @classmethod
    def of(cls, detail: SetDetail) -> EvaluationSetDetailResponse:
        return cls(
            **_set_fields(detail.set),
            counts=ItemCountsResponse.of(detail.counts),
            items=[EvaluationItemResponse.of(item) for item in detail.items],
        )


class EvaluationSetList(BaseModel):
    items: list[EvaluationSetResponse]


class EvaluationSetRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=MAX_SET_NAME_LENGTH)
    description: str | None = Field(default=None, max_length=2000)


class EvaluationSetPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=MAX_SET_NAME_LENGTH)
    description: str | None = Field(default=None, max_length=2000)


class ImportRequest(BaseModel):
    """A window of the gateway's own log. ``uncited`` narrows it the way the monitoring
    filter does: ``true`` for requests whose answer cited nothing, ``false`` for the ones
    that did, omitted for both."""

    model_config = ConfigDict(extra="forbid")

    start: datetime = Field(alias="from")
    end: datetime = Field(alias="to")
    uncited: bool | None = None
    limit: int = Field(default=200, ge=1, le=500)


class ImportResponse(BaseModel):
    imported: int
    duplicates: int
    skipped: int
    labelled: int

    @classmethod
    def of(cls, result: ImportResult) -> ImportResponse:
        return cls(
            imported=result.imported,
            duplicates=result.duplicates,
            skipped=result.skipped,
            labelled=result.labelled,
        )


class GenerateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    count: int = Field(default=10, ge=1, le=25)


class GenerateResponse(BaseModel):
    generated: int
    failed: int
    tokens_in: int
    tokens_out: int
    model_name: str | None

    @classmethod
    def of(cls, result: GenerateResult) -> GenerateResponse:
        return cls(
            generated=result.generated,
            failed=result.failed,
            tokens_in=result.tokens_in,
            tokens_out=result.tokens_out,
            model_name=result.model_name,
        )


class RunRequest(BaseModel):
    """``memory_config`` is the editor's unsaved form, merged the way Try retrieval merges
    it; omit it to run the saved gateway."""

    model_config = ConfigDict(extra="forbid")

    memory_config: dict[str, Any] | None = None


class EvaluationRunList(BaseModel):
    items: list[EvaluationRunSummaryResponse]


class MetricDeltaResponse(BaseModel):
    name: str
    before: float | None
    after: float | None
    change: float | None


class RunDiffResponse(BaseModel):
    before: EvaluationRunSummaryResponse
    after: EvaluationRunSummaryResponse
    metrics: list[MetricDeltaResponse]
    config_changes: dict[str, list[Any]]
    index_changes: list[str]
    won: list[dict[str, Any]]
    lost: list[dict[str, Any]]

    @classmethod
    def of(cls, diff: RunDiff) -> RunDiffResponse:
        return cls(
            before=EvaluationRunSummaryResponse.of(diff.before),
            after=EvaluationRunSummaryResponse.of(diff.after),
            metrics=[
                MetricDeltaResponse(
                    name=delta.name, before=delta.before, after=delta.after, change=delta.change
                )
                for delta in diff.metrics
            ],
            config_changes={key: list(value) for key, value in diff.config_changes.items()},
            index_changes=list(diff.index_changes),
            won=[dict(entry) for entry in diff.won],
            lost=[dict(entry) for entry in diff.lost],
        )


def _set_fields(row: EvaluationSet) -> dict[str, Any]:
    return {
        "id": row.id,
        "gateway_id": row.gateway_id,
        "name": row.name,
        "description": row.description,
        "created_at": row.created_at,
        "updated_at": row.updated_at,
    }


def _optional(value: Any) -> str | None:
    return str(value) if value not in (None, "") else None


__all__ = [
    "AuditAlertList",
    "AuditAlertResponse",
    "AuditRequest",
    "AuditResponse",
    "AuditStatusResponse",
    "ChunkingReportResponse",
    "EmbeddingReportResponse",
    "EvaluationItemPatch",
    "EvaluationItemRequest",
    "EvaluationItemResponse",
    "EvaluationRunList",
    "EvaluationRunResponse",
    "EvaluationRunSummaryResponse",
    "EvaluationSetDetailResponse",
    "EvaluationSetList",
    "EvaluationSetPatch",
    "EvaluationSetRequest",
    "EvaluationSetResponse",
    "FindingResponse",
    "GenerateRequest",
    "GenerateResponse",
    "ImportRequest",
    "ImportResponse",
    "LabelRequest",
    "RunDiffResponse",
    "RunRequest",
]

File: web/src/pages/validation.ts
/**
 * The arithmetic and the sentences behind the Validation screens (task 103), kept out of
 * the components so a table of inputs can pin them.
 *
 * Two families. The **audit** helpers turn a stored report into what the screen says
 * about it — its age, its worst finding, where a finding's fix lives — and rebuild the
 * server's histogram buckets client-side for Compare, which has the chunks in hand and no
 * report. The **evaluation** helpers format the headline numbers the same way everywhere
 * (a recall of `null` is "—", never "0.00"), and write the one sentence a run's headline
 * has to carry when its labels were not all a person's.
 */

import type { Tone } from '@/components/status'
import type {
  AuditResponse,
  AuditStatusResponse,
  EvaluationItemResponse,
  EvaluationRunSummaryResponse,
  EvaluationSetResponse,
  FindingResponse,
  HistogramResponse,
} from '@/api/types'

export type AuditKind = 'chunking' | 'embedding'

export const AUDIT_KINDS: readonly { kind: AuditKind; label: string }[] = [
  { kind: 'chunking', label: 'Chunking' },
  { kind: 'embedding', label: 'Embeddings' },
]

//  The server's own bucketing: twelve buckets across [0, 1.5 × chunk_size], the last
//  one open. Mirrored here rather than requested, so Compare's histogram and the audit's
//  draw on the same axis for the same setting.
const HISTOGRAM_BUCKETS = 12
const OVER_CEILING_FACTOR = 1.5

// -- audits -----------------------------------------------------------------

export function severityTone(severity: string | null | undefined): Tone {
  if (severity === 'red') return 'error'
  if (severity === 'amber') return 'warn'
  if (severity === 'green') return 'ok'
  return 'neutral'
}

export function severityLabel(severity: string | null | undefined): string {
  if (severity === 'red') return 'needs attention'
  if (severity === 'amber') return 'worth a look'
  if (severity === 'green') return 'healthy'
  return 'not audited'
}

/** "2 hours ago", from the audit's finish (or start, while it runs). */
export function auditAge(audit: AuditResponse | null | undefined, now = Date.now()): string {
  if (!audit) return 'Never run.'
  const when = new Date(audit.finished_at ?? audit.created_at).getTime()
  const minutes = Math.max(0, Math.round((now - when) / 60_000))
  if (audit.status === 'running') return 'Running now.'
  const age =
    minutes < 1
      ? 'just now'
      : minutes < 60
        ? `${minutes} min ago`
        : minutes < 60 * 48
          ? `${Math.round(minutes / 60)} h ago`
          : `${Math.round(minutes / (60 * 24))} days ago`
  return audit.status === 'failed' ? `Failed ${age}.` : `Run ${age}.`
}

/** The sentence beside the Run button: what the audit found, in one line. */
export function auditHeadline(audit: AuditResponse | null | undefined): string {
  if (!audit || !audit.report) return ''
  const red = audit.report.findings.filter((f) => f.severity === 'red').length
  const amber = audit.report.findings.filter((f) => f.severity === 'amber').length
  const points = `${audit.report.points.toLocaleString()} points`
  if (red === 0 && amber === 0) return `${points}, nothing to report.`
  const parts = []
  if (red) parts.push(`${red} red`)
  if (amber) parts.push(`${amber} amber`)
  return `${points}, ${parts.join(' and ')} finding${red + amber === 1 ? '' : 's'}.`
}

/**
 * Where a finding's fix lives, as a link. A badly cut file opens Compare with the
 * document preselected; a mixed index opens the connector's reindex; a model problem
 * opens the platform's; the rest open the document itself.
 */
export function findingLink(
  connectorId: string,
  finding: FindingResponse,
  documentId?: string,
): { to: string; label: string } | null {
  if (finding.action === 'compare' && documentId) {
    return {
      to: `/connectors/${connectorId}?compare=${documentId}`,
      label: 'Open in Compare',
    }
  }
  if (finding.action === 'reindex') {
    return {
      to: `/connectors/${connectorId}#chunking`,
      label: 'Reindex this connector',
    }
  }
  if (finding.action === 'platform') {
    return { to: '/platform/settings', label: 'Platform embedding settings' }
  }
  if (documentId) {
    return {
      to: `/connectors/${connectorId}?document=${documentId}`,
      label: 'Open document',
    }
  }
  return null
}

export function driftCostLine(status: AuditStatusResponse | undefined): string {
  if (!status) return ''
  const { sample, tokens, points } = status.drift_estimate
  if (points === 0) return 'Nothing is indexed to re-embed.'
  return `Re-embeds ${sample.toLocaleString()} of ${points.toLocaleString()} chunks now — about ${tokens.toLocaleString()} tokens at the embedding provider.`
}

/**
 * Token counts into the server's buckets. Used by Compare, which has chunks rather than
 * a report; the audit renders the report's own buckets, computed the same way.
 */
export function chunkHistogram(
  tokenCounts: readonly number[],
  chunkSize: number,
): HistogramResponse {
  const width = Math.max(1, Math.ceil((chunkSize * OVER_CEILING_FACTOR) / HISTOGRAM_BUCKETS))
  const counts = new Array<number>(HISTOGRAM_BUCKETS).fill(0)
  for (const tokens of tokenCounts) {
    const index = Math.min(HISTOGRAM_BUCKETS - 1, Math.floor(tokens / width))
    counts[index] = (counts[index] ?? 0) + 1
  }
  return {
    bucket_tokens: width,
    buckets: counts.map((count, index) => ({
      lower: index * width,
      upper: (index + 1) * width,
      count,
    })),
  }
}

// -- evaluation -------------------------------------------------------------

export function metric(value: number | null | undefined, digits = 2): string {
  return value === null || value === undefined ? '—' : value.toFixed(digits)
}

/** Reads a nested number out of a run's metrics without trusting its shape. */
export function dig(metrics: Record<string, unknown>, ...path: string[]): number | null {
  let value: unknown = metrics
  for (const key of path) {
    if (!value || typeof value !== 'object') return null
    value = (value as Record<string, unknown>)[key]
  }
  return typeof value === 'number' ? value : null
}

/** "recall@6 0.82 · precision@6 0.41 · MRR 0.77" — the chunk column before the budget. */
export function runHeadline(run: EvaluationRunSummaryResponse): string {
  if (run.status !== 'succeeded') return runStatusLine(run)
  const k = dig(run.metrics, 'k') ?? 0
  return [
    `recall@${k} ${metric(dig(run.metrics, 'all', 'chunk', 'recall'))}`,
    `precision@${k} ${metric(dig(run.metrics, 'all', 'chunk', 'precision'))}`,
    `MRR ${metric(dig(run.metrics, 'all', 'chunk', 'mrr'))}`,
  ].join(' · ')
}

export function runStatusLine(run: EvaluationRunSummaryResponse): string {
  if (run.status === 'queued') return 'Queued.'
  if (run.status === 'running') {
    return `Running: ${run.completed_items} of ${run.total_items} questions.`
  }
  if (run.status === 'failed') return `Failed: ${run.error ?? 'no reason recorded'}.`
  return ''
}

export function runTone(status: string): Tone {
  if (status === 'succeeded') return 'ok'
  if (status === 'failed') return 'error'
  return 'info'
}

/** The warnings the server put on the run's headline, if any. */
export function runWarnings(run: EvaluationRunSummaryResponse): string[] {
  const warnings = (run.metrics as { warnings?: unknown }).warnings
  return Array.isArray(warnings) ? warnings.map(String) : []
}

export function sourceLabel(source: EvaluationItemResponse['source']): string {
  return {
    log: 'from the log',
    citation: 'from a citation',
    manual: 'by hand',
    generated: 'by a model',
  }[source]
}

/** The list row's second line: what a set holds, and what that does to its numbers. */
export function setSummary(set: EvaluationSetResponse): string {
  const { total = 0, verified = 0, generated = 0, negatives = 0 } = set.counts ?? {}
  if (total === 0) return 'No questions yet.'
  const parts = [`${total} question${total === 1 ? '' : 's'}`]
  parts.push(`${verified} verified`)
  if (negatives) parts.push(`${negatives} negative`)
  if (generated) parts.push(`${generated} written by a model`)
  return parts.join(' · ')
}

/** What pressing Run will spend: one embedding call per question. */
export function runCostLine(set: EvaluationSetResponse | { counts: { total: number } }): string {
  const total = Math.min(set.counts.total, 500)
  if (total === 0) return 'Nothing to run.'
  return `${total} embedding call${total === 1 ? '' : 's'} — one per question. Nothing is sent to a completion model and nothing is written to the index.`
}

export function formatDelta(change: number | null): string {
  if (change === null) return '—'
  if (Math.abs(change) < 0.005) return '±0.00'
  return `${change > 0 ? '+' : ''}${change.toFixed(2)}`
}

export function deltaTone(change: number | null, higherIsBetter = true): Tone {
  if (change === null || Math.abs(change) < 0.005) return 'neutral'
  return change > 0 === higherIsBetter ? 'ok' : 'error'
}

/** The default import window: the last seven days, as the request body wants them. */
export function lastWeek(now = new Date()): { from: string; to: string } {
  const to = new Date(now)
  const from = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000)
  return { from: from.toISOString(), to: to.toISOString() }
}
